src/libraries/System.Private.CoreLib/src/System/Globalization/InvariantModeCasing.cs

```diff
@@ -0,0 +1,779 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Globalization
+{
+    internal static partial class InvariantModeCasing
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToLower(char c)
+        {
+           Debug.Assert(GlobalizationMode.Invariant);
+
+           ushort v = LowerCase1[c >> 8];
+           v = LowerCase2[v + ((c >> 4) & 0xF)];
+           v = LowerCase3[v + (c & 0xF)];
+
+           return v == 0 ? c : (char)v;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToUpper(char c)
+        {
+           Debug.Assert(GlobalizationMode.Invariant);
+
+           ushort v = UpperCase1[c >> 8];
+           v = UpperCase2[v + ((c >> 4) & 0xF)];
+           v = UpperCase3[v + (c & 0xF)];
+
+           return v == 0 ? c : (char)v;
+        }
+
+        internal static unsafe string ToLower(string s)
+        {
+            if (s.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            fixed (char* pSource = s)
+            {
+                int i = 0;
+                while (i < s.Length)
+                {
+                    if (char.IsHighSurrogate(pSource[i]) && i < s.Length - 1 && char.IsLowSurrogate(pSource[i + 1]))
+                    {
+                        SurrogateCasing.ToLower(pSource[i], pSource[i + 1], out ushort h, out ushort l);
+                        if (pSource[i] != h || pSource[i + 1] != l)
+                        {
+                            break;
+                        }
+
+                        i += 2;
+                        continue;
+                    }
+
+                    if (ToLower(pSource[i]) != pSource[i])
+                    {
+                        break;
+                    }
+
+                    i++;
+                }
+
+                if (i >= s.Length)
+                {
+                    return s;
+                }
+
+                string result = string.FastAllocateString(s.Length);
+                fixed (char* pResult = result)
+                {
+                    for (int j = 0; j < i; j++)
+                    {
+                        pResult[j] = pSource[j];
+                    }
+
+                    Span<char> resultSpan = new Span<char>(pResult, s.Length).Slice(i);
+                    InvariantModeCasing.ToLower(s.AsSpan().Slice(i), resultSpan);
+                }
```

Comment:
I don't have a strong opinion, as this is corelib and in the heart of globalization, but should we avoid proliferating using pointers to write into FastAllocateStrings, and instead use String.Create? e.g.
```C#
return string.Create(s.Length, (s, i), static (destination, state) =>
{
    ReadOnlySpan<char> source = state.s;
    source.Slice(0, state.i).CopyTo(destination);
    InvariantModeCasing.ToLower(source.Slice(state.i), destination.Slice(state.i));
});
```

Reply:
good idea. I'll try that.